source/opt/spread_volatile_semantics.cpp

```diff
@@ -0,0 +1,143 @@
+
+namespace spvtools {
+namespace opt {
+namespace {
+
+const uint32_t kOpDecorateInOperandBuiltinDecoration = 2u;
+const uint32_t kOpLoadInOperandMemoryOperands = 1u;
+
+bool IsBuiltInForVolatileSemantics(uint32_t built_in) {
+  switch (built_in) {
+    case SpvBuiltInSMIDNV:
+    case SpvBuiltInWarpIDNV:
+    case SpvBuiltInSubgroupSize:
+    case SpvBuiltInSubgroupLocalInvocationId:
+    case SpvBuiltInSubgroupEqMask:
+    case SpvBuiltInSubgroupGeMask:
+    case SpvBuiltInSubgroupGtMask:
+    case SpvBuiltInSubgroupLeMask:
+    case SpvBuiltInSubgroupLtMask:
+      return true;
+    default:
+      return false;
+  }
+}
+
+bool HasVolatileDecoration(analysis::DecorationManager* decoration_manager,
+                           uint32_t var_id) {
+  return decoration_manager->HasDecoration(var_id, SpvDecorationVolatile);
+}
+
+}  // namespace
+
+Pass::Status SpreadVolatileSemantics::Process() {
+  Status status = Status::SuccessWithoutChange;
+  bool is_vk_memory_model_enabled = context()->get_feature_mgr()->HasCapability(
+      SpvCapabilityVulkanMemoryModel);
+  for (Instruction& var : context()->types_values()) {
+    if (!IsTargetForVolatileSemantics(&var)) {
+      continue;
+    }
+
+    if (is_vk_memory_model_enabled) {
+      SetVolatileForLoads(&var);
+    } else {
+      DecorateVarWithVolatile(&var);
+    }
+    status = Status::SuccessWithChange;
+  }
+  return status;
+}
+
+void SpreadVolatileSemantics::DecorateVarWithVolatile(Instruction* var) {
+  analysis::DecorationManager* decoration_manager =
+      context()->get_decoration_mgr();
+  uint32_t var_id = var->result_id();
+  if (HasVolatileDecoration(decoration_manager, var_id)) {
+    return;
+  }
+  get_decoration_mgr()->AddDecoration(
+      SpvOpDecorate,
+      {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {var_id}},
+       {spv_operand_type_t::SPV_OPERAND_TYPE_ID, {SpvDecorationVolatile}}});
+}
+
+void SpreadVolatileSemantics::SetVolatileForLoads(Instruction* var) {
+  std::vector<Instruction*> updated_insts;
+  auto* def_use_mgr = context()->get_def_use_mgr();
+  def_use_mgr->ForEachUser(var, [&updated_insts](Instruction* user) {
+    if (user->opcode() != SpvOpLoad) {
+      return;
+    }
+
+    updated_insts.push_back(user);
+    if (user->NumInOperands() <= kOpLoadInOperandMemoryOperands) {
+      user->AddOperand({SPV_OPERAND_TYPE_ID, {SpvMemoryAccessVolatileMask}});
+      return;
+    }
+
+    uint32_t memory_operands =
+        user->GetSingleWordInOperand(kOpLoadInOperandMemoryOperands);
+    memory_operands |= SpvMemoryAccessVolatileMask;
+    user->SetInOperand(kOpLoadInOperandMemoryOperands, {memory_operands});
+  });
+
+  for (auto* inst : updated_insts) {
+    def_use_mgr->AnalyzeInstUse(inst);
+  }
+}
+
+bool SpreadVolatileSemantics::IsTargetForVolatileSemantics(Instruction* var) {
+  analysis::DecorationManager* decoration_manager =
+      context()->get_decoration_mgr();
+  auto execution_model = context()->GetExecutionModel();
```

Comment:
If there are multiple execution models in the module this method return SpvExecutionModelMax. In that case, no changes will be made.

Since all inputs must be listed in the entry point, perhaps you should instead just iterate through entry points and check their IO for the appropriate builtins. Just keep a visited set in case the builtin is used by multiple entry points.

Reply:
@jaebaek Do you have reason to believe that a single execution model will be used for these cases?  When the code for `GetExecutionModel` was originally written, the only tools generating spir-v for shaders had a single entry point.  Also, if a function was called from multiple entry points, it could be legal from one entry point but not the other.  This was a reasonable simplifying assumption.

However, does ray tracing generate multiple entry points per spir-v file?  Is it possible that two entry points use the same variable and it is a target for one, but not the other?  I don't know the spec on this point.

Reply:
@jaebaek Do you have an answer for my second question?  Is it possible that two entry points use the same variable and it is a target for one, but not the other?  This is an important question.  

If the answer is yes, then you will have to identify the cases where you cannot do the transformation.  You'll have to return an error with a message telling them to inline a particular function or something like that.

If the answer is no, then you do not need the execution model to know if the variable is a target, and you can make the code much simpler.